### PR TITLE
[F] Product + Content Refresh Refactor (ENT-2207)

### PR DESCRIPTION
**Pull request by Ceiu**
Several changes from the PoC, and more to come. Feel free to give suggestions on the implementation of the node builders, processors (visitors), and mappers. Those are still somewhat malleable at this point and I'm open to suggestions on improvements.

I think the biggest gotcha here is that the visitors are not stateless as one may expect. I don't think it's a huge issue, but it's abnormal from what I've seen in the past.

### CODE REVIEW

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+
+        node.markVisited();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void complete() {
+        this.ownerProductCurator.saveAll(this.ownerProductEntities, false, true);
```
Comment (nikosmoum) [inline server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java:153]: The ProductManager did flush when saving the core products, then again when saving the owner-products. Here it seems we're deferring flushing for when? Shouldn't we flush before CandlepinPoolManager starts processing the pools? (same question applies for `ContentNodeVisitor.complete`)

File: server/src/main/java/org/candlepin/controller/refresher/RefreshWorker.java [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+import java.util.Collection;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+
+
+/**
+ * The RefreshWorker gathers upstream objects to refresh, and then performs the actual work to
+ * update their local representations.
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/RefreshWorker.java:54]: The overall algorithm for merging doesn't change, just how we're doing the work. I'm not entirely opposed to documenting how/why this works, but from an outsider's perspective, nothing (should have) changed here.

File: server/src/main/java/org/candlepin/controller/refresher/builders/NodeFactory.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        if (mapper == null) {
+            throw new IllegalArgumentException("mapper is null");
+        }
+
+        this.mapper = mapper;
+        return this;
+    }
+
+    /**
+     * Returns an entity node for the given entity class and ID, creating it as necessary.
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/builders/NodeFactory.java:91]: Added an extra bit to the end of that sentence. Now reads "creating and mapping it as necessary". Will show up in the next commit.

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java (hunk 1 above)
Comment (nikosmoum) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java:153]: yes, evict to false would be more clear

File: server/src/main/java/org/candlepin/controller/refresher/RefreshWorker.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikosmoum) [reply server/src/main/java/org/candlepin/controller/refresher/RefreshWorker.java:54]: I'm thinking more as a resource for developers than outsiders

File: server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java (hunk 1 above)
Comment (Ceiu) [reply server/src/main/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitor.java:153]: Done.

File: server/src/test/java/org/candlepin/controller/PoolManagerTest.java [hunk 4]
```diff
@@ -545,75 +556,98 @@ private void mockProducts(Owner owner, Product... products) {
...
-        }
-
-        this.mockContentImport(owner, contentMap);
-    }
+                return output;
+            }})
+        .when(this.refreshWorker).execute(eq(owner));
+    }
+
+    // private void mockProductImport(Owner owner, final Map<String, Product> products) {
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/test/java/org/candlepin/controller/PoolManagerTest.java:582]: Yeah. Once I finish with the refresh testing, I'm going to go back and remove the commented out code which is no longer required, and fix up the bits that are.

File: server/src/test/java/org/candlepin/controller/refresher/builders/ContentNodeBuilderTest.java [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+     *  the ID of the entity represented by the new entity node
+     *
+     * @throws IllegalStateException
+     *  if an entity node representing the given ID cannot be created
+     *
+     * @return
+     *  a new EntityNode instance representing the entity with the provided ID
+     */
+    // EntityNode<E, I> buildNode(NodeFactory factory, Owner owner, String id);
+
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/src/test/java/org/candlepin/controller/refresher/builders/ContentNodeBuilderTest.java:116]: Oh, whoops. Something I copied in so I didn't have to keep switching tabs. Removed.

File: server/src/test/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitorTest.java [hunk 6]
```diff
@@ -332,8 +277,10 @@ public void testProcessNodeForUnchangedEntity(String key, Object base, Object up
         Owner owner = TestUtil.createOwner();
         String id = TestUtil.randomString("test_id");
 
-        Content existing = this.createPopulatedEntity(id, key, base);
-        ContentInfo imported = this.createPopulatedEntity(id, key, base);
+        Content existing = this.createPopulatedExistingEntity(id, key, base);
+        ContentInfo imported = this.createPopulatedImportedEntity(id, key, base);
+
+        MergeValidator<ContentInfo> validator = this.validators.get(key);
```
Comment (nikosmoum) [inline server/src/test/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitorTest.java:283]: It doesn't look like this one line is needed (the validator isn't used thereafter)

File: server/src/test/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitorTest.java [hunk 7]
```diff
@@ -60,79 +55,9 @@
...
-        }
-
-        @Override
-        Collection access(ContentInfo content);
-    }
-
-    private Map<String, Mutator> mutators;
-    private Map<String, Validator> validators;
+    private Map<String, Mutator<Content>> contentMutators;
+    private Map<String, Mutator<ContentInfo>> cinfoMutators;
```
Comment (nikosmoum) [inline server/src/test/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitorTest.java:59]: :+1:

File: server/src/test/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitorTest.java [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+    @ParameterizedTest(name = "{displayName} {index}: {0}")
+    @MethodSource("productDataProvider")
+    public void testProcessNodeForUnchangedEntity(String key, Object base, Object update) {
+        Owner owner = TestUtil.createOwner();
+        String id = TestUtil.randomString("test_id");
+
+        Product existing = this.createPopulatedExistingEntity(id, key, base);
+        ProductInfo imported = this.createPopulatedImportedEntity(id, key, base);
+
+        MergeValidator<ProductInfo> validator = this.validators.get(key);
```
Comment (nikosmoum) [inline server/src/test/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitorTest.java:273]: Same here, validator not used after retrieving here

File: server/src/test/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitorTest.java (hunk 6 above)
Comment (Ceiu) [reply server/src/test/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitorTest.java:283]: Fixed

File: server/src/test/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitorTest.java (hunk 8 above)
Comment (Ceiu) [reply server/src/test/java/org/candlepin/controller/refresher/visitors/ProductNodeVisitorTest.java:273]: Fixed

File: server/bin/upstream_subscription_injector.rb [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+owner_key = ARGV.shift
+
+if owners.empty? || !owners.include?(owner_key)
+  log("INFO", "Owner \"#{owner_key}\" does not exist; creating it")
+  candlepin.create_owner(owner_key)
+end
+
+# Ensure that the subscription data file was specified and exists
+terminate("A subscription data file must be specified") if ARGV.empty?
```
[the remark this answers is not in the document]
Comment (Ceiu) [reply server/bin/upstream_subscription_injector.rb:180]: Fixed

**Pull request merged**